### ISSUE TITLE
Camera perspective yfov upper limit

### ISSUE THREAD

**Issue by lexaknyazev** [KhronosGroup/glTF#1806]
_Originally posted by @vpenades [here](https://github.com/KhronosGroup/glTF-Validator/issues/140)._
> The official specification states that the value must be over 0, and although it does not explicitly state the upper limit, it states that the value is defined as radians, which sets an implicit upper limit.
>
> Also, the spec says the value must be used in formulas using the form `Tan(0.5*yfov)` which implicitly sets the upper limit below π.
>
> I guess the spec does not define the upper limit because it's a non rational number? but math libraries certainly define it...

I think it's a spec overlook. Given that camera properties go straight into pre-defined equations, there's no reason to allow values greater than or equal to π. We have similar restriction for punctual spot lights.

@donmccurdy @bghgary 
Do you think it's possible to change the schema? At the very least, we need to update the description.

**Comment by bghgary** [comment]
> Do you think it's possible to change the schema?

Not normally. What happens when the values are greater than π?

**Comment by lexaknyazev** [comment]
Reportedly, some math libraries do not like such values. Given that `yfov` is used as `tan(yfov / 2)`, the following happens:
- `[0]` - undefined (already forbidden)
- `(0, π)` - normal behavior
- `[π]` - undefined
- `(π, 2π)` - flipped frustrum
- `[2π]` - undefined
- `(2π, 3π)` - normal behavior
- (continue forever)

![image](https://user-images.githubusercontent.com/3479527/80995253-083d0a00-8e4f-11ea-899b-7eeda82f55b3.png)

**Comment by bghgary** [comment]
I think we can update the description and add checks for validator. Schema changes seem too dangerous at this point (though I don't see anything specific will break).

**Comment by lexaknyazev** [comment]
@donmccurdy @pjcozzi @emackey 
Any opinions here?

**Comment by vpenades** [comment]
given that most math libraries work in 32bit floating point, and fov ranges very close to π don't make much sense, what about setting the upper limit to (float32)π - (float32)epsilon ?

**Comment by lexaknyazev** [comment]
@vpenades 
That would be an implementation detail - some platforms (such as ECMAScript) may still use float64. I wouldn't expect consistent rendering results with any `yfov` values close to π as the final results would be unstable anyway. Values such as `1.5 * π` are likely a bigger issue.

The spec question here is whether we're able to update the schema, or just the description.

**Comment by donmccurdy** [comment]
No preference on this.

If the 2.0 schema doesn't allow the change, a 2.1 release should allow it, correct? I feel like we are starting to have a critical mass of smaller fixes that could justify a 2.1...

**Comment by lexaknyazev** [comment]
Strictly following semver, this would be a breaking change thus requiring 3.0 release.

By keeping the schema as-is, are we saying that it's OK to use those values even if the results make little sense? Given that the spec provides exact matrices, it doesn't look like an undefined behavior, even with intermediate Infinity values.

**Comment by bghgary** [comment]
> Strictly following semver, this would be a breaking change thus requiring 3.0 release.

Since we are talking about versioning, I just want to call out [this section](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#versioning) in the spec which has some text on minor versions.

Just curious now, even if we could change the schema, how does one put a max value of π with JSON schema? How many digits of π would you use?

> By keeping the schema as-is, are we saying that it's OK to use those values even if the results make little sense?

I think it would be fine to add some verbiage stating the range in the spec. I don't think there will be any serious compatibility issues. The only thing I can think of is if there are lots of models already out there with values that are outside of this range, but I suspect that's not true.

**Comment by donmccurdy** [comment]
Ok, I do understand that we can't change the 2.0 schema in a way that causes 2.0 assets to no longer validate. But I'm not sure I understand why making the specification more strict would be a breaking change in a minor semver release. Consider this compatibility matrix:

| - | 2.0 asset | 2.1 asset |
|---|---|---|
| 2.0 client | ✅ | ✅ |
| 2.1 client | ❓  | ✅ |

Is the problem that a new client implementing 2.1 might have to do additional work to support a 2.0 asset? Or that not all 2.0 assets are automatically valid 2.1 assets? If not this, then I don't know what would be valid use of a minor semver release. I haven't been able to find a crisp definition of how exactly semver should be understood in the context of file formats, though.



> I think it would be fine to add some verbiage stating the range in the spec. I don't think there will be any serious compatibility issues.

+1

**Comment by donmccurdy** [comment]
> Backwards compatibility will ensure that any client implementation that supports loading a glTF 2.x asset will also be able to load a glTF 2.0 asset. 

Ah I see, thanks for the link @bghgary. We could perhaps bend this rule given that no client can do anything "correct" with these values anyway, within the spec, but I won't insist on it.

**Comment by vpenades** [comment]
Regarding backwards compatibility:

Cameras are probably some of the least used features of the whole glTF schema. So if a model has a camera it's probably because it does use the camera. Now, if the camera has an over π value, it will render nothing in the best case, and will make the client crash in the worst case. So I can't see why being backwards compatible to a crash can be useful.

> @lexaknyazev Values such as `1.5 * π` are likely a bigger issue.

Indeed, because I've found a glTF where the yfov was set in degrees. I am using a library that calculates the projection matrix from the camera parameters, and if the values are not withing range, that is, __yfov less than π__, it throws an exception. It should not be the case if all parameters are pre-validated... and that's how I was aware of the issue.

If you're curious, the system library checking the limits is System.Numerics.Vectors, in [this line](https://github.com/dotnet/runtime/blob/e64bc548c609455652fcd4107f1f4a2ac3084ff3/src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs#L862).

My particular problem is that I'm generating code from the schema, and since the max value validation is missing, I have to hack into my code generation to include it. It's done now, so my library already checks for invalid yfov max values _regardless of the schema semver_, so strictly speaking, I've already broke compatibility with ver 2 glTFs that would crash my lib. No big loss.

Finally, I really don't understand why is this such a big deal, I mean, we've been breaking backwards compatibility _all the time_, not by adding things to the schema, but by _reinterpreting_ parts of the schema. it comes to mind the reinterpretation of the skin weights to sum __exactly__ one. This reinterpretation broke like 80% of sketchfab models alone!

As I said, this is not a big deal, it's a very very minor feature and I have it fixed it in my lib. If I reported this issue is because I was concerned other implementations might not be aware of this issue, and because someone could make an exploit by intentionally authoring a glTF with invalid values.

I don't think adding a clarification would be very useful, it's a feature so minor that would probably be overlooked by most developers writing implementations.
